DEV_GUIDE.md

```diff
@@ -0,0 +1,145 @@
+      --ngpus 1 \
+      --minbytes 128 \
+      --maxbytes 16G \
+      --stepbytes 1M \
+      --op sum \
+      --datatype float \
+      --root 0 \
+      --iters 100 \
+      --warmup_iters 50 \
+      --agg_iters 1 \
+      --average 1 \
+      --parallel_init 0 \
+      --check 1 \
+      --blocking 0 \
+      --cudagraph 0 \
+      --stepfactor 2
+   ```
+
+## Setting Up the Test Template in CloudAI
+### Directory Structure and Expected Files
+Create a new directory under `cloudai/schema/test_template` for organizing the new test template. This structure typically includes various Python modules to define the operational parameters of the test within CloudAI:
+   ```plaintext
+   $ tree cloudai/schema/test_template
+   cloudai/schema/test_template
+   ├── nccl_test
+   │   ├── __init__.py
+   │   ├── grading_strategy.py
+   │   ├── command_gen_strategy.py
+   │   ├── install_strategy.py
+   │   └── template.py
+   ```
+
+### Implementation of Strategies
+1. **Template Class**
+
+   Define the `NcclTest` class to enable CloudAI to recognize and instantiate this as a test template. You can declare variables used within the template.
+   ```python
+   class NcclTest(TestTemplate):
+       SUPPORTED_SUBTESTS = ["bisection_perf_mpi", ...]
+   ```
+
+2. **Install Strategy**
+
+    Implement an installation strategy to handle the setup of necessary components on the system where the test will run. Decorators in Python are used here to register the strategy with a strategy registry. This approach allows CloudAI to dynamically recognize and use the appropriate strategy based on the test template and system configuration.
+   ```python
+   @StrategyRegistry.strategy(InstallStrategy, [SlurmSystem], [NcclTest])
+   class NcclTestSlurmInstallStrategy(SlurmInstallStrategy):
+       def install(self):
+           # Logic to install the test template
+       def uninstall(self):
+           # Logic to uninstall the test template
+   ```
+
+3. **Command Generation Strategy**
+
+   Develop a command generation strategy that formulates the commands needed to execute:
+   ```python
+   @StrategyRegistry.strategy(CommandGenStrategy, [SlurmSystem], [NcclTest])
+   class NcclTestSlurmCommandGenStrategy(SlurmCommandGenStrategy):
+       def gen_exec_command(self, env_vars, cmd_args):
+           # Command to execute the test template
+   ```
+
+4. **cloudai/schema/test_template/\_\_init\_\_.py**
+
+    Update \_\_init\_\_.py accordingly.
+    ```python
+    ....
+    from .nccl_test import (
+        NcclTest,
+        NcclTestGradingStrategy,
+        NcclTestReportGenerationStrategy,
+        NcclTestSlurmCommandGenStrategy,
+        NcclTestSlurmInstallStrategy,
+    )
+    ...
+
+    __all__ = [
+        ...
+        "NcclTest",
+        "NcclTestGradingStrategy",
+        "NcclTestReportGenerationStrategy",
+        "NcclTestSlurmCommandGenStrategy",
+        "NcclTestSlurmInstallStrategy",
+        ...
+    ]
+    ```
+
+### Testing Procedures
```

Comment:
Do you think we can make it a `pytest`?

Reply:
Maybe I am not experienced enough to set it up as a pytest. I have several concerns regarding testing these with pytest. Installation involves interaction with the file system and remote docker repositories. Running involves interaction with the cluster. Therefore, with pytest, it must be challenging to validate. Others will work with pytest.